mempool/src/pool.rs

```diff
@@ -0,0 +1,2768 @@
+
+        // Now that the minimum rolling fee has been bumped up, a low-fee tx will not pass
+        // validation
+        let child_2 = tx_spend_input(
+            &mempool,
+            TxInput::new(outpoint_source_id.clone(), 2, DUMMY_WITNESS_MSG.to_vec()),
+            None,
+            flags,
+            locktime,
+        )?;
+        log::debug!("before child2");
+        assert!(matches!(
+            mempool.add_transaction(child_2),
+            Err(Error::TxValidationError(
+                TxValidationError::RollingFeeThresholdNotMet { .. }
+            ))
+        ));
+
+        // We provide a sufficient fee for the tx to pass the minimum rolling fee requirement
+        let child_2_high_fee = tx_spend_input(
+            &mempool,
+            TxInput::new(outpoint_source_id, 2, DUMMY_WITNESS_MSG.to_vec()),
+            mempool.get_minimum_rolling_fee().compute_fee(estimate_tx_size(1, 1))?,
+            flags,
+            locktime,
+        )?;
+        log::debug!("before child2_high_fee");
+        mempool.add_transaction(child_2_high_fee)?;
+
+        // We simulate a block being accepted so the rolling fee will begin to decay
+        mempool.process_block(&parent_id)?;
+
+        // Because the rolling fee is only updated when we attempt to add a tx to the mempool
+        // we need to submit a "dummy" tx to trigger these updates.
+
+        // Since memory usage is now zero, it is less than 1/4 of the max size
+        // and ROLLING_FEE_BASE_HALFLIFE / 4 is the time it will take for the fee to halve
+        let halflife = ROLLING_FEE_BASE_HALFLIFE / 4;
+        mock_clock.increment(halflife);
+        let dummy_tx = TxGenerator::new().generate_tx(&mempool)?;
+        log::debug!("first attempt to add dummy");
+        assert!(matches!(
+            mempool.add_transaction(dummy_tx.clone()),
+            Err(Error::TxValidationError(
+                TxValidationError::RollingFeeThresholdNotMet { .. }
+            ))
+        ));
+        assert_eq!(
+            mempool.get_minimum_rolling_fee(),
+            (rolling_fee / FeeRate::new(2)).unwrap()
+        );
+
+        mock_clock.increment(halflife);
+        // Fee will have dropped under INCREMENTAL_RELAY_FEE_RATE / 2 by now, so it will be set to
+        // zero and our tx will be submitted successfully
+        log::debug!("second attempt to add dummy");
+        mempool.add_transaction(dummy_tx)?;
+        assert_eq!(mempool.get_minimum_rolling_fee(), FeeRate::new(0));
+        Ok(())
+    }
+
+    #[test]
+    fn different_size_txs() -> anyhow::Result<()> {
+        use std::time::Duration;
+        use std::time::Instant;
+        let mut mempool = setup();
+        let initial_tx = TxGenerator::new()
+            .with_num_inputs(1)
+            .with_num_outputs(10_000)
+            .generate_tx(&mempool)?;
+        mempool.add_transaction(initial_tx)?;
+
+        let target_txs = 100;
+        let mut time_processing_txs = Duration::from_millis(0);
```

Comment:
We don't need this still right? This was just when you were working out the bottleneck

Reply:
Yeah, we can do away with the time measurements, but I'm guessing we still want to keep the test right? Since we wanted a test with different-sized txs...

Reply:
Yeah keep the tests but bin the timing

Reply:
Done!

Reply:
Done.